internal/expr/bindinputs.go

```diff
@@ -49,41 +35,111 @@ func (te *TypedExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 		if v.Kind() != reflect.Struct && v.Kind() != reflect.Map {
 			return nil, fmt.Errorf("need struct or map, got %s", t.Kind())
 		}
-		if _, ok := typeValue[t]; ok {
+		if _, ok := typeToValue[t]; ok {
 			return nil, fmt.Errorf("type %q provided more than once", t.Name())
 		}
-		typeValue[t] = v
-		typeNames = append(typeNames, t.Name())
-		if !inQuery[t] {
-			// Check if we have a type with the same name from a different package.
-			for _, typeMember := range te.inputs {
-				if t.Name() == typeMember.OuterType().Name() {
-					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", typeMember.OuterType().String(), t.String())
+		typeToValue[t] = v
+	}
+
+	// Generate SQL and query parameters.
+	params := []any{}
+	outputs := []typeinfo.Member{}
+	argTypeUsed := map[reflect.Type]bool{}
+	inCount := 0
+	outCount := 0
+	sqlStr := bytes.Buffer{}
+	for _, te := range *tes {
+		switch te := te.(type) {
+		case *typedInputExpr:
+			typeMember := te.input
+			outerType := typeMember.OuterType()
+			v, ok := typeToValue[outerType]
+			if !ok {
+				return nil, inputMissingError(outerType, typeToValue)
+			}
+			argTypeUsed[outerType] = true
+
+			val, err := typeMember.ValueFromOuter(v)
+			if err != nil {
+				return nil, err
+			}
+			params = append(params, sql.Named("sqlair_"+strconv.Itoa(inCount), val.Interface()))
+
+			sqlStr.WriteString("@sqlair_" + strconv.Itoa(inCount))
+			inCount++
+		case *typedOutputExpr:
+			for i, oc := range te.outputColumns {
+				sqlStr.WriteString(oc.sql)
+				sqlStr.WriteString(" AS ")
+				sqlStr.WriteString(markerName(outCount))
+				if i != len(te.outputColumns)-1 {
+					sqlStr.WriteString(", ")
 				}
+				outCount++
+				outputs = append(outputs, oc.tm)
 			}
-			return nil, fmt.Errorf("%s not referenced in query", t.Name())
+		case *bypass:
+			sqlStr.WriteString(te.chunk)
 		}
 	}
 
-	// Query parameters.
-	var params []any
-	for i, typeMember := range te.inputs {
-		outerType := typeMember.OuterType()
-		v, ok := typeValue[outerType]
-		if !ok {
-			return nil, typeMissingError(outerType.Name(), typeNames)
+	for argType := range typeToValue {
+		if !argTypeUsed[argType] {
+			return nil, fmt.Errorf("%s not referenced in query", argType.Name())
 		}
+	}
 
-		val, err := typeMember.ValueFromOuter(v)
-		if err != nil {
-			return nil, err
-		}
+	return &PrimedQuery{outputs: outputs, sql: sqlStr.String(), params: params}, nil
+}
 
-		params = append(params, sql.Named("sqlair_"+strconv.Itoa(i), val.Interface()))
+// typedExpression represents a expression bound to a type. It contains
+// information to generate the SQL for the part and to access Go types
+// referenced in the part.
```

Comment:
I've changed the first sentence as suggested. For the second sentence I've actually removed it and put a modified version on `TypeBoundExpr` itself because I think it belongs more there. The sentence that was there before was a bit vague. We now have:
```
// TypeBoundExpr represents a SQLair query bound to concrete Go types. It
// contains information used to generate the underlying SQL query and map it to
// the SQLair query.

// typedExpression represents a expression with the type names bound to Go
// types.
```

Reply:
I love the new comments, they are crystal clear to me

Reply:
My only comment would be against using the word "query" as it seems to be a heated topic.

Reply:
Changed to statement.